lightningd/bolt11.c

```diff
@@ -0,0 +1,1002 @@
+#include <bitcoin/address.h>
+#include <bitcoin/chainparams.h>
+#include <bitcoin/script.h>
+#include <ccan/array_size/array_size.h>
+#include <ccan/cast/cast.h>
+#include <ccan/endian/endian.h>
+#include <ccan/structeq/structeq.h>
+#include <ccan/tal/str/str.h>
+#include <common/utils.h>
+#include <errno.h>
+#include <hsmd/gen_hsm_wire.h>
+#include <inttypes.h>
+#include <lightningd/bech32.h>
+#include <lightningd/bolt11.h>
+#include <lightningd/hsm_control.h>
+#include <lightningd/lightningd.h>
+#include <lightningd/log.h>
+#include <secp256k1_recovery.h>
+#include <stdarg.h>
+#include <stdio.h>
+#include <wire/wire.h>
+#include <wire/wire_sync.h>
+
+/* 1000 * 10^8 millisatoshi == 1 bitcoin */
+#define MSAT_PER_BTC 100000000000ULL
+
+struct multiplier {
+        const char letter;
+        /* We can't represent p postfix to msat, so we multiply this by 10 */
+        u64 m10;
+};
+
+/* BOLT #11:
+ *
+ * The following `multiplier` letters are defined:
+ *
+ * * `m` (milli): multiply by 0.001
+ * * `u` (micro): multiply by 0.000001
+ * * `n` (nano): multiply by 0.000000001
+ * * `p` (pico): multiply by 0.000000000001
+ */
+static struct multiplier multipliers[] = {
+        { 'm', 10 * MSAT_PER_BTC / 1000 },
+        { 'u', 10 * MSAT_PER_BTC / 1000000 },
+        { 'n', 10 * MSAT_PER_BTC / 1000000000 },
+        { 'p', 10 * MSAT_PER_BTC / 1000000000000ULL }
+};
+
+/* If pad is false, we discard any bits which don't fit in the last byte.
+ * Otherwise we add an extra byte */
+static bool pull_bits(struct hash_u5 *hu5,
+                      u5 **data, size_t *data_len, void *dst, size_t nbits,
+                      bool pad)
+{
+        size_t n5 = nbits / 5;
+        size_t len = 0;
+
+        if (nbits % 5)
+                n5++;
+
+        if (*data_len < n5)
+                return false;
+        if (!bech32_convert_bits(dst, &len, 8, *data, n5, 5, pad))
+                return false;
+        if (hu5)
+                hash_u5(hu5, *data, n5);
+        *data += n5;
+        *data_len -= n5;
+
+        return true;
+}
+
+/* For pulling fields where we should have checked it will succeed already. */
+#ifndef NDEBUG
+#define pull_bits_certain(hu5, data, data_len, dst, nbits, pad)             \
```

Comment:
Shouldn't there be a semicolon before the `assert`?

Reply:
? We're literally replacing x_certain(...) with assert(x(...)).

Reply:
Aww, now I see, was confusing the replacee with the replacement :-)